planner/core/rule_partition_processor.go

```diff
@@ -113,6 +115,101 @@ func generateHashPartitionExpr(t table.Table, ctx sessionctx.Context, columns []
 	return exprs[0], nil
 }
 
+func (s *partitionProcessor) findUsedPartitions(ds *DataSource, pi *model.PartitionInfo) ([]int, error) {
+	pe, err := generateHashPartitionExpr(ds.table, ds.ctx, ds.TblCols, ds.names)
+	if err != nil {
+		return nil, err
+	}
+	partIdx := expression.ExtractColumns(pe)
+	colLen := make([]int, 0, len(partIdx))
+	for i := 0; i < len(partIdx); i++ {
+		partIdx[i].Index = i
+		colLen = append(colLen, types.UnspecifiedLength)
+	}
+	datcher, err := ranger.DetachCondAndBuildRangeForIndex(ds.SCtx(), ds.allConds, partIdx, colLen)
+	if err != nil {
+		return nil, err
+	}
+	ranges := datcher.Ranges
+	used := make([]int, 0, len(ranges))
+	for _, r := range ranges {
+		if r.IsPointNullable(ds.ctx.GetSessionVars().StmtCtx) {
+			if !r.HighVal[0].IsNull() {
+				if len(r.HighVal) != len(partIdx) {
+					used = []int{-1}
+					break
+				}
+			}
+			pos, isNull, err := pe.EvalInt(ds.ctx, chunk.MutRowFromDatums(r.HighVal).ToRow())
+			if err != nil {
+				return nil, err
+			}
+			if isNull {
+				pos = 0
+			}
+			used = append(used, int(math.Abs(pos%int64(pi.Num))))
+		} else {
+			used = []int{-1}
+			break
+		}
+	}
+	return used, nil
+}
+
+func (s *partitionProcessor) produceHashPartitionPlan(ds *DataSource, used []int, pi *model.PartitionInfo) (LogicalPlan, error) {
+	if len(used) > 0 && used[0] == -1 {
+		return s.makeUnionAllChildren(ds, pi, fullRange(len(pi.Definitions)))
+	}
+	children := make([]LogicalPlan, 0, len(pi.Definitions))
+	for _, pos := range used {
+		if len(ds.partitionNames) > 0 && !s.findByName(ds.partitionNames, pi.Definitions[pos].Name.L) {
+			// For condition like `from t partition (p1) where a = 5`, but they are conflict, return TableDual directly.
```

Comment:
`used` is an array, if **all** of the used partitions are not in the selection, return table dual

`from t partition (p1) where a = 5 or a = 10` ...

Reply:
fixed, currently i use `makeUnionAllChildren` and remove `convertToRangeOr` to convert `used` into `partitionRangeOR`